### PR TITLE
Supporting logits as parameters in Bernoulli and Categorical

### PR DESCRIPTION
**Pull request by neerajprad**
This allows for using logits directly as parameters in Bernoulli and Categorical distributions. 

**Details:**
 - Uses the binary cross entropy operation to score Bernoulli samples. This makes the `log_prob` computation differentiable.
 - Categorical normalizes the probs/logits parameters provided in the constructor and stores these as attributes, so that other methods do not need to bother with normalization.
 - Simplifies entropy computation (and corrects it for the case of Categorical with unnormalized probability values as parameters).

### CODE REVIEW

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import torch
 from torch.autograd import Variable
 from torch.distributions.distribution import Distribution
-from torch.distributions.utils import broadcast_all
+from torch.distributions.utils import broadcast_all, get_logits, get_probs
```
Comment (fritzo) [inline torch/distributions/bernoulli.py]: nit: Consider naming these (`probs_to_logits`, `logits_to_probs`) or (`get_logits_from_probs`, `get_probs_from_logits`).

File: torch/distributions/utils.py [hunk 2]
```diff
@@ -75,3 +76,57 @@ def broadcast_all(*values):
...
+    """
+    if not isinstance(tensor, Variable):
+        return F.softmax(Variable(tensor), -1).data
+    return F.softmax(tensor, -1)
+
+
+def log_sum_exp(tensor):
+    """
+    Numerically stable implementation for the `LogSumExp` operation. The
+    summing is done along the last dimension.
```
Comment (fritzo) [inline torch/distributions/utils.py]: nit: Consider either mentioning that the last dimension is kept or adding a `keepdim=False` argument so that it is clear in usage `log_sum_exp(tensor, keepdim=True)` what the output shape is.

File: torch/distributions/bernoulli.py [hunk 3]
```diff
@@ -37,20 +45,17 @@ def sample(self, sample_shape=torch.Size()):
         shape = self._extended_shape(sample_shape)
         return torch.bernoulli(self.probs.expand(shape))
 
+    def _binary_cross_entropy(self, value):
```
Comment (fritzo) [inline torch/distributions/bernoulli.py]: :+1: Thanks for implementing this! Dustin will be happy.

**Comment by neerajprad** [comment]
Thanks for the review, @fritzo. Addressed your comments, and opened up a PR in [pytorch/pytorch#4448](https://github.com/pytorch/pytorch/pull/4448)
